File: app/src/main/java/weatherapp/tittojose/me/weatherapp/ui/presenter/HomeScreenPresenter.java
package weatherapp.tittojose.me.weatherapp.ui.presenter;

import weatherapp.tittojose.me.weatherapp.model.pojo.WeatherModel;
import weatherapp.tittojose.me.weatherapp.repository.WeatherRepositoryContract;
import weatherapp.tittojose.me.weatherapp.ui.view.HomeScreen;

public class HomeScreenPresenter implements HomePresenterContract {

    private final WeatherRepositoryContract.WeatherLoadListener weatherLoadListener;
    private HomeScreen homeScreen;
    private WeatherRepositoryContract weatherRepository;


    public HomeScreenPresenter(final HomeScreen homeScreen, WeatherRepositoryContract weatherRepository) {
        this.homeScreen = homeScreen;
        this.weatherRepository = weatherRepository;
        weatherLoadListener = new WeatherRepositoryContract.WeatherLoadListener() {
            @Override
            public void onWeatherDataLoadedSuccess(WeatherModel weatherModel) {
                homeScreen.hideLoading();
                homeScreen.onWeatherDataLoadSuccess(weatherModel);
            }

            @Override
            public void onWeatherDataLoadedFailed() {
                homeScreen.hideLoading();
                homeScreen.onWeatherDataLoadFailed();
            }
        };
        this.weatherRepository.setWeatherLoadListener(weatherLoadListener);
    }


    @Override
    public void loadWeatherData() {
        this.homeScreen.showLoading();
        this.weatherRepository.getWeatherData();
    }

    @Override
    public void retryLoadWeatherData() {
        loadWeatherData();
    }

}

File: app/src/main/java/weatherapp/tittojose/me/weatherapp/presenter/HomeScreenPresenter.kt
package weatherapp.tittojose.me.weatherapp.presenter

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import weatherapp.tittojose.me.weatherapp.model.WeatherAPI
import weatherapp.tittojose.me.weatherapp.view.HomeScreen

class HomeScreenPresenter(
	private val homeScreenView: HomeScreen,
	private val weatherApi: WeatherAPI
) {
	private val disposables: CompositeDisposable = CompositeDisposable()

	fun loadWeatherData() {
		val disposable = weatherApi.getCurrentWeather()
			.subscribeOn(Schedulers.io())
			.observeOn(AndroidSchedulers.mainThread())
			.doOnSubscribe { homeScreenView.showLoading() }
			.subscribe({
				homeScreenView.hideLoading()
				homeScreenView.onWeatherDataLoadSuccess(it)
			}, {
				homeScreenView.hideLoading()
				homeScreenView.onWeatherDataLoadFailed()
			})
		disposables.add(disposable)
	}

	fun retryLoadWeatherData() {
		loadWeatherData()
	}

	fun onDestroy() {
		disposables.dispose()
	}
}